x/shield/keeper/rewards.go

```diff
@@ -58,21 +60,60 @@ func (k Keeper) PayoutNativeRewards(ctx sdk.Context, addr sdk.AccAddress) (sdk.C
 	return ctkRewards, nil
 }
 
+// getServiceFeesForBlock calculates service fees for current block.
+func (k Keeper) getServiceFeesForBlock(ctx sdk.Context) types.MixedDecCoins {
+	totalServiceFees := k.GetServiceFees(ctx)
+	// currentBlockSecond - previousBlockSecond
+	secondsFromLastDistribution := sdk.NewDecFromInt(sdk.NewInt(int64(ctx.BlockTime().Sub(ctx.WithBlockHeight(ctx.BlockHeight() - 1).BlockTime()).Seconds())))
+	// (currentBlockSecond - previousBlockSecond) * totalServiceFees / protectionPeriodSeconds
+	serviceFees := totalServiceFees.MulDec(secondsFromLastDistribution).QuoDec(sdk.NewDecFromInt(sdk.NewInt(int64(k.GetPoolParams(ctx).ProtectionPeriod.Seconds()))))
+	return serviceFees
+}
+
 // DistributeFees distributes service fees to all providers.
-func (k Keeper) DistributeFees(ctx sdk.Context) {
-	serviceFees := k.GetServiceFees(ctx)
+func (k Keeper) distributeFees(ctx sdk.Context, serviceFees types.MixedDecCoins) {
+	serviceFeesLeft := k.GetServiceFeesLeft(ctx)
 	totalCollateral := k.GetTotalCollateral(ctx)
-
 	providers := k.GetAllProviders(ctx)
 	for _, provider := range providers {
-		proportion := sdk.NewDecFromInt(sdk.MaxInt(provider.Collateral, sdk.ZeroInt())).QuoInt(totalCollateral)
-		nativeFees := serviceFees.Native.MulDecTruncate(proportion)
-
-		serviceFees.Native = serviceFees.Native.Sub(nativeFees)
-
+		// fees * providerCollateral / totalCollateral
+		nativeFees := serviceFees.Native.MulDec(sdk.NewDecFromInt(provider.Collateral).QuoInt(totalCollateral))
 		provider.Rewards = provider.Rewards.Add(types.MixedDecCoins{Native: nativeFees})
 		k.SetProvider(ctx, provider.Address, provider)
+
+		serviceFeesLeft.Native = serviceFeesLeft.Native.Sub(nativeFees)
 	}
+	k.SetServiceFeesLeft(ctx, serviceFeesLeft)
+}
 
-	k.SetServiceFees(ctx, serviceFees)
+// UpdateServiceFees update service fees based on purchases' completion time.
+func (k Keeper) updateServiceFees(ctx sdk.Context) {
```

Comment:
Why do we still need this function at all?

Reply:
When we call removeExpiredPurchases(), we update service fees for purchases with
`previousBlockTime < ProtectionTime <= DeletionTime <= currentBlockTime`.
This function updates service fees for purchases with
`previousBlockTime < ProtectionTime <= currentBlockTime <= DeletionTime`